[core] Enable dogfood

adangel opened the pull request:
This enables PMD on PMD with a first rule. It will break the build, if a violation is found.
As suggested in #361, we'll do an incremental approach and the [ruleset](https://github.com/pmd/build-tools/blob/master/config/src/main/resources/net/sourceforge/pmd/pmd-dogfood-config.xml) only contains a single rule for now.

Note: Since I added now a snapshot dependency to pmd-build-tools, we'll need to release [build-tools](https://github.com/pmd/build-tools) before we release PMD 6.1.0.

The only rule for now is UnusedPrivateField.
It seems, we need to improve this rule to consider cases like in JavaFX, see #907

jsotuyod commented:
@adangel nice! You are absolutely right about the `@FXML` FP...  come to think of it, we probably also have FPs for things annotated as `@Deprecated`.

adangel commented:
I've now enabled a couple of other rules and fixed the issues so far.
The ruleset currently in use is this: https://github.com/pmd/build-tools/blob/master/config/src/main/resources/net/sourceforge/pmd/pmd-dogfood-config.xml

I've added all rules, that are not enabled, as comments to get a better overview, where we are...

I've rebased this PR to the current master. I think, it can be merged now as is.
I'm sorry, it became a bit big...

oowekyala commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/Helper.java:
```diff
@@ -211,11 +211,16 @@ static String getFQVariableName(final ASTNewKeyValueObjectExpression variable) {
         for (TypeRef intObject : interfaces) {
-            return isWhitelisted(intObject.getNames());
+            if (isWhitelisted(intObject.getNames())) {
+                hasWhitelistedInterfaces = true;
```
`return true` would be cleaner... or am I missing the point of the rule ?

adangel replied:
yes, it's not really clear. If I remember correctly, the rule that triggered, was [AvoidBranchingStatementAsLastInLoop](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_errorprone.html#avoidbranchingstatementaslastinloop). And looking now again at the original code - that's actually a bug - it would have only checked the first interface of the list.

Yes, I think, using just `return true` here and avoiding the local variable should work...

oowekyala commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/ForLoopCanBeForeachRule.java:
```diff
@@ -395,16 +396,19 @@ private boolean isReplaceableIteratorLoop(Entry<VariableNameDeclaration, List<Na
-                && ((indexName + ".hasNext").equals(image) || (indexName + ".next").equals(image))) {
-                continue;
+            if (!isCallingNext) {
+                onlyCallingNext = false;
```
same here

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/benchmark/BenchmarkResult.java:
```diff
@@ -40,7 +40,13 @@ public int compareTo(BenchmarkResult benchmarkResult) {
         if (cmp == 0) {
             long delta = this.time - benchmarkResult.time;
-            cmp = delta > 0 ? 1 : (delta < 0 ? -1 : 0);
+            if (delta > 0) {
```
`cmp = Integer.compare(delta, 0)` ?

adangel replied:
nice!

oowekyala commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/CycloBaseVisitor.java:
```diff
@@ -25,12 +25,12 @@
  * @see net.sourceforge.pmd.lang.java.metrics.impl.CycloMetric
  */
-public class CycloBaseVisitor extends JavaParserControllessVisitorAdapter {
+public final class CycloBaseVisitor extends JavaParserControllessVisitorAdapter {
```
Visitors from this package were originally intended to be extended by users to make their custom metrics. It's even [documented on the website](https://pmd.github.io/pmd-6.1.0/pmd_userdocs_extending_metrics_howto.html#best-practices). Any reasons why we may want to change that?

adangel replied:
Good catch! The rule that triggered here as [ClassWithOnlyPrivateConstructorsShouldBeFinal](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_design.html#classwithonlyprivateconstructorsshouldbefinal).
Then I'll make the default constructors of these visitor classes protected (otherwise you can't subclass them, if there is no accessible constructor....)

adangel replied:
Or I'll remove the (empty) constructor alltogether - similar to what we have in [AtfdBaseVisitor](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/visitors/AtfdBaseVisitor.java) and we get the default constructor back.

oowekyala commented:
I'm merging that tonight, unless @jsotuyod you want some more time to review?

oowekyala commented:
@adangel Merged, congrats :)
